### PR TITLE
observability: Log error messages from Protobuf handlers

### PR DESCRIPTION
**Pull request by AlexAPB**
This fixes an issue where all errors returned from a Protobuf handler
were logged uninformatively (identical to Thrift exceptions):

```
{ ... { ... "error": "application_error"}}
```

Unfortunately, all errors returned from Protobuf handlers are marked
as application errors and changing this behavior would be a breaking
change to HTTP/JSONpb users :(

https://github.com/yarpc/yarpc-go/blob/bd70ffbd17e635243988ba62be97eebff738204d/encoding/protobuf/inbound.go#L71

With this PR we

1. Correctly log the error using `zap.Error(err)`, by detecting if we
   observed an error in addition to the application error bit being
   set.

2. Any `yarpcerror` created with details (including
  `encoding/protobuf.WithDetails`) are logged at the application error
  log level. This aligns with Thrift exceptions and would be the least
  surprising behavior for users (perhaps auto) migrating from Thrift
  exceptions to Protobuf error details.

Note that we made a similar workaround for metrics in #1739, where
regardless of the applciation bit being set or not, we'd emit the
correct metric.

Fixes T5590765

- [x] Description and context for reviewers: one partner, one stranger
- [x] Entry in CHANGELOG.md

### CODE REVIEW

File: internal/observability/middleware_test.go [hunk 1]
```diff
@@ -205,7 +207,7 @@ func TestMiddlewareLogging(t *testing.T) {
 
 	tests := []test{
 		{
-			desc:            "no downstream errors",
+			desc:            "sucess",
```
Comment (mh-park) [inline internal/observability/middleware_test.go:210]: typo: sucess -> success

File: internal/observability/call.go [hunk 2]
```diff
@@ -97,9 +99,22 @@ func (c call) endLogs(elapsed time.Duration, err error, isApplicationError bool)
 		}
 
 		lvl := c.levels.failure
-		if isApplicationError {
+
+		// Application errors are
+		//  - Thrift exceptions
+		//  - `yarpcerror`s with error details
```
[the remark this answers is not in the document]
Comment (AlexAPB) [reply internal/observability/call.go:105]: Gotcha, I can see why this is confusing. This comment is referring to what we categorize as an "application logs" and the next comment is referring to marking the application error bit. Let me try to clarify this further

**Comment by AlexAPB** [comment]
> So, in Thrift, if someone returns a non-Thrift-exception error (including a YARPC error), it's propagated up without being classified as an application error. I know it's difficult to have Protobuf and Thrift be on the same page, but the behavior is essentially different. Are we completely OK with that?

For me, in short, yes.

Getting Protobuf/Thrift on the same page would require us to make a breaking change for
- HTTP/JSONpb consumers, which unfortunately are difficult to enumerate; and
- outbound middleware checking the `transport.Response#ApplicationError`

The idea of an "application error" is only in the APIs to support TChannel. In a Post TChannel/Thrift world of only gRPC/Protobuf, this concept is effectively meaningless.

That being said, this change pivots to the least surprising behavior for users, making migrations from Thrift to gRPC/Protobuf more straightforward.

- Thrift exceptions are logged at application level and so the equivalent in Protobuf, (error details) should be too. (This is especially useful for auto-migrations, as some service owners turn off logging for Thrift exceptions since they're "too noisy"/expected).
- Errors messages returned in Thrift handlers (not exceptions) are correctly logged and so any error from Protobuf handlers should be logged too.

I'd argue that this change offers more convergence and [less surprising behavior](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). Let me know your thoughts!

File: internal/observability/call.go [hunk 3]
```diff
@@ -97,9 +99,22 @@ func (c call) endLogs(elapsed time.Duration, err error, isApplicationError bool)
...
+		//
+		// Any error returned from a Protobuf handler is marked as an application
+		// error (unlike Thrift). Therefore, we distinguish an application error
+		// from a regular error by inspecting if an error detail was set.
```
[the remark this answers is not in the document]
Comment (AlexAPB) [reply internal/observability/call.go]: updated

File: internal/observability/call.go [hunk 4]
```diff
@@ -97,9 +99,25 @@ func (c call) endLogs(elapsed time.Duration, err error, isApplicationError bool)
...
+		// Thrift exceptions to Protobuf error details.
+		//
+		// Unfortunately, all errors returned from a Protobuf handler are marked as
+		// an application error on the 'transport.ResponseWriter'. Therefore, we
+		// distinguish an application error from a regular error by inspecting if an
+		// error detail was set.
+		//
+		// https://github.com/yarpc/yarpc-go/pull/1912
+		hasErrDetails := len(yarpcerrors.FromError(err).Details()) > 0
+		if appErrBitWithNoError || (isApplicationError && hasErrDetails) {
```
Comment (AllenLuUber) [inline internal/observability/call.go:117]: nit:
```
if hasErrDetails := len(yarpcerrors.FromError(err).Details()) > 0; appErrBitWithNoError || (isApplicationError && hasErrDetails)  {
}
```

File: internal/observability/call.go [hunk 5]
```diff
@@ -83,6 +83,8 @@ func (c call) EndWithPanic(err error) {
 }
 
 func (c call) endLogs(elapsed time.Duration, err error, isApplicationError bool) {
+	appErrBitWithNoError := isApplicationError && err == nil // ie Thrift exception
```
Comment (AllenLuUber) [inline internal/observability/call.go:86]: let's be a bit more specific here in the comments. 
```
// true if success or a thrift error
appErrBitWithNoError := isApplicationError && err == nil
```

Comment (AlexAPB) [reply internal/observability/call.go:86]: Actually, this case is only for Thrift exceptions, not successes.

**Pull request merged**